Move files from os-common

chaitu236 opened the pull request:
Move nisetreboottype, xfce files from os-common

@ni/rtos

chaitu236 committed:
xfce-nilrt-settings: Move xserver-xfce to OE

Add changes in xserver-xfce from os-common to OE

Signed-off-by: Chaitanya Vadrevu <chaitanya.vadrevu@ni.com>

chaitu236 committed:
initscripts-nilrt: Move nisetreboottype into OE

Move nisetreboottype from os-common to OE

Signed-off-by: Chaitanya Vadrevu <chaitanya.vadrevu@ni.com>

chaitu236 committed:
initscripts-nilrt: Move modules_autoload into OE

Move modules_autoload from os-common into OE

Signed-off-by: Chaitanya Vadrevu <chaitanya.vadrevu@ni.com>

chaitu236 committed:
initscripts-nilrt: Update nisetupkernelconfig

Update nisetupkernelconfig with latest changes from os-common

Signed-off-by: Chaitanya Vadrevu <chaitanya.vadrevu@ni.com>

chaitu236 committed:
initscripts-nilrt: Move nisetupdump into OE

Move nisetupdump from os-common into OE

Signed-off-by: Chaitanya Vadrevu <chaitanya.vadrevu@ni.com>

shruthi-ravi commented on recipes-xfce/xfce-nilrt-settings/xfce-nilrt-settings_1.0.bb:
```diff
@@ -65,6 +66,7 @@ FILES_${PN} = "${confdir}/autostart/dpms.desktop \
 	    ${confdir}/xfce4/xfconf/xfce-perchannel-xml/xfwm4.xml \
 	    ${confdir}/xfce4/xfconf/xfce-perchannel-xml/xsettings.xml \
 	    /usr/local/natinst/bin/dual-monitor-setup.sh \
+	    /etc/default/xserver-xfce \
```
Here's another [recipe](https://github.com/ni/meta-nilrt/blob/nilrt/master/sumo/recipes-xfce/xserver-xfce-init/xserver-xfce-init_1.0.bb#L27) that adds this file. I think you can merge the contents of this default file to that file. I think that would be the part where we read from the nirtcfg settings to check if UI is enabled.

amstewart commented on recipes-ni/initscripts-nilrt/files/mountconfig:
```diff
@@ -0,0 +1,58 @@
+		then
+			# If an mtd partition has this partition label, use
+			# mount_ubi_volume
+			mount_ubi_volume "$partition_label" "$volume_label" \
+				"$volume_number" "$mount_point" "$mount_args"
```
The `config` partition only exists on ARM targets, iirc. So I don't think this initscript is relevant to dunfell images and beyond. It can be dropped.

amstewart commented on recipes-ni/initscripts-nilrt/files/nicreatelogdir:
```diff
@@ -0,0 +1,11 @@
+# Ensure the log directory is always present with the correct
+# ownership.
+
+log_dir="/var/local/natinst/log"
+
+mkdir -p -m 777 $log_dir
+chown lvuser:ni $log_dir
```
Do we really need an initscript for this? Surely there is either (a) already another initscript which creates NI-owned directories or (b) we can do what every other distro does and expect that log-creating packages will make their own directories.

chaitu236 replied:
It does seem excessive to have an initscript just for this. But I [haven't been able to find](https://codesearch.natinst.com/search/text?q=mkdir.*natin%20repo%3Ameta-nilrt&fold_case=false&regex=true&context=true) another initscript that creates NI dirs.

As for (b), I suspect the reason this was added was to support some old piece of software that didn't creates its own dir. Do you think we should remove this to find all such cases and have their devs fix them?

amstewart replied:
>Do you think we should remove this to find all such cases and have their devs fix them?

It [seems like](https://codesearch.natinst.com/search/text?q=local%2Fnatinst%2Flog&fold_case=auto&regex=true&context=true) that path is used in a lot of different packages. In all cases, it seems like the intention is just to give the `lvuser` user a location to access core dumps, LV crash logs, ect. (which implies that it is where LV customers are intended to store their logs).

It's probably too risky to just remove it from the base image and require the NIFeeds package owners to fix all their packages. But I might like to move this code to something like the [`niacctbase`](https://github.com/ni/meta-nilrt/blob/nilrt/master/dunfell/recipes-core/niacctbase/niacctbase.bb) package, which does generic setup for the `lvuser` account.

shruthi-ravi replied:
I found a [recipe](https://github.com/ni/meta-nilrt/blob/nilrt/master/dunfell/recipes-support/syslog-ng/syslog-ng_3.%25.bbappend#L25) that creates the directory too.

amstewart commented on recipes-ni/initscripts-nilrt/files/nisetreboottype:
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+# The purpose of this script is to set the reboot type of the controller based
+# on the value of ni_boot_mode_input. Reboots to install mode require a cold
+# reboot. All other modes should use the default reboot type
```
I don't know if this is a relevant workflow; it seems like the remnants of the obsolete "install mode" workflow. Can you check if SCS or anyone else even sets `/sys/kernel/ni_requested_reboot_type` to `install`? If not, we can drop this whole file.

chaitu236 replied:
Found [this](https://p4.natinst.com/browser/perforce/MAX/components/remote/softwareinstall/trunk/21.0/source/SoftwareActionShared/Imaging.cpp?revision_at=51501451#line_388)... so keep this around?

amstewart replied:
I... don't know what that's supposed to mean? Does that codepath run when `reboot_type` is set to `install`?

chaitu236 replied:
The codepath runs when ni_boot_mode_input is set to install and [this](https://p4.natinst.com/browser/perforce/MAX/components/remote/softwareinstall/trunk/21.0/source/SoftwareActionShared/Imaging.cpp?revision_at=51501451#line_388) does that.

shruthi-ravi commented on recipes-ni/initscripts-nilrt/files/mountconfig:
```diff
@@ -0,0 +1,58 @@
+	if ! grep -q " $mount_point " /proc/self/mounts; then
+		if cat /sys/class/mtd/mtd*/name 2>/dev/null | grep -q ^$partition_label\$
+		then
+			# If an mtd partition has this partition label, use
```
The entries that refer to mtd and mount_ubi_volume are Zynq-specific and can be removed. The mount command on line 33 is enough. Do you think it's a good idea to edit these files as you move them or should it be done in the next update? Can you create a task for it or something if you pick the latter option?

chaitu236 replied:
Removed the file altogether as remaining changes aren't necessary. I can edit them as I go.

amstewart commented on recipes-ni/initscripts-nilrt/files/nisetupdump:
```diff
@@ -0,0 +1,18 @@
+if [ -f $safemodeflag ]; then
+	echo "/mnt/userfs/var/local/natinst/log/core_dump.%E" > /proc/sys/kernel/core_pattern
+else
+	echo "/var/local/natinst/log/core_dump.%E" > /proc/sys/kernel/core_pattern
+fi
+
+#Enable core dump for tainted binaries
+echo 1 > /proc/sys/fs/suid_dumpable
```
We need to keep these settings*, but the rest of this script is only relevant to safemode. Surely we already have a package in OE that sets `ulimit` or something? That's probably a better place to put sysctl options like this.

* Though the `core_pattern` should be set to a better location. Where should that be @gratian ?

chaitu236 replied:
There is [this](https://github.com/ni/meta-nilrt/blob/nilrt/master/sumo/recipes-ni/base-files-nilrt/files/ulimit.sh#L1) but it gets installed to /etc/profile.d, which is probably not where we want to add this.
How about [nisetupkernelconfig](https://github.com/ni/meta-nilrt/blob/nilrt/master/sumo/recipes-ni/initscripts-nilrt/files/nisetupkernelconfig#L1)?

chaitu236 replied:
Adding core_pattern also to nisetupkernelconfig for now

gratian replied:
Yeah, I think moving the core_pattern  into nisetupkernelconfig is OK.

shruthi-ravi commented on recipes-ni/initscripts-nilrt/initscripts-nilrt_1.0.bb:
```diff
@@ -78,6 +79,8 @@ do_install () {
 	install -m 0755 ${WORKDIR}/mountconfig           ${D}${sysconfdir}/init.d
 	# script for auto-loading kernel modules (needed for modules that are not auto-loaded by udev)
 	install -m 0755 ${WORKDIR}/modules_autoload      ${D}${sysconfdir}/init.d
+	# nicreatelogdir script that ensures log directory is present with proper ownership
```
Description feels redundant.

amstewart commented on recipes-ni/initscripts-nilrt/initscripts-nilrt_1.0.bb:
```diff
@@ -85,10 +98,16 @@ do_install () {
 	update-rc.d -r ${D} nicreatecpuacctgroups start 2  4 5 .
 	update-rc.d -r ${D} nipopulateconfigdir   start 36 S .
-	update-rc.d -r ${D} nisetupkernelconfig   start 3  4 5 .
+	update-rc.d -r ${D} nisetupkernelconfig   start 3  5 .
```
This script will eventually need to be on RL 4 as well, but we can do that in the later RAUC story to configure RL 4 as the new "safemode".

shruthi-ravi commented on recipes-ni/initscripts-nilrt/initscripts-nilrt_1.0.bb:
```diff
@@ -78,6 +79,8 @@ do_install () {
 	# script for auto-loading kernel modules (needed for modules that are not auto-loaded by udev)
 	install -m 0755 ${WORKDIR}/modules_autoload      ${D}${sysconfdir}/init.d
+	# nicreatelogdir script that ensures log directory is present with proper ownership
+	install -m 0755 ${WORKDIR}/nicreatelogdir        ${D}${sysconfdir}/init.d
```
This [recipe](https://github.com/ni/meta-nilrt/blob/nilrt/master/dunfell/recipes-support/syslog-ng/syslog-ng_3.%25.bbappend#L25) seems to be creating the same directory.

shruthi-ravi commented on recipes-ni/initscripts-nilrt/initscripts-nilrt_1.0.bb:
```diff
@@ -75,6 +76,8 @@ do_install () {
 	install -m 0755 ${WORKDIR}/wirelesssetdomain     ${D}${sysconfdir}/init.d
 	install -m 0755 ${WORKDIR}/cleanvarcache         ${D}${sysconfdir}/init.d
 	install -m 0755 ${WORKDIR}/mountconfig           ${D}${sysconfdir}/init.d
+	# script for auto-loading kernel modules (needed for modules that are not auto-loaded by udev)
```
I think it would be more appropriate to move this description to the script.

amstewart commented on recipes-xfce/xfce-nilrt-settings/files/xserver-xfce.default:
```diff
@@ -0,0 +1,10 @@
+	exit 0
+fi
+
+XUSER=lvuser
+HOME=/home/lvuser
```
Seems like this file already exists in [xserver-xfce-init-1.0](https://github.com/ni/meta-nilrt/blob/29fbbd49b52e84de6fb1b3c18702b743d2124420/recipes-xfce/xserver-xfce-init/xserver-xfce-init-1.0/xserver-xfce.conf) and is installed to the same place [here](https://github.com/ni/meta-nilrt/blob/29fbbd49b52e84de6fb1b3c18702b743d2124420/recipes-xfce/xserver-xfce-init/xserver-xfce-init_1.0.bb#L27).

shruthi-ravi reviewed (changes requested):
Part 1..I still have the last 3 commits to look at.

shruthi-ravi commented:
Well, looks like I was a minute late :) @amstewart has similar concerns.

shruthi-ravi commented on recipes-ni/initscripts-nilrt/files/nisetupkernelconfig:
```diff
@@ -16,7 +16,16 @@ do
+# not until the next time those threads are scheduled. We want to remove the possibility
+# of that happening during jitter-sensitive code in a user application.
+PIDS=`grep -l irq/ /proc/*/comm 2>/dev/null | cut -d/ -f3`
+if [ -z "$PIDS" ] ; then
```
Should this really be an error? I don't understand why. This seems to be the only difference between existing code and new code, other than the comment itself?

amstewart replied:
Yeah; I agree that this doesn't imply an error.

chaitu236 replied:
Should I change the msg to "No IRQ threads found" and exit 0? [Ref](https://review-board.natinst.com/r/138441/)

chaitu236 replied:
@billpittman What do you think?

chaitu236 replied:
Ok I just tried the original and don't see an error when there are no irq threads - so not sure why we added it.

shruthi-ravi replied:
Ok, I think we can keep the original file with the updated comment unless @gratian objects.

gratian replied:
I'll defer to @billpittman since he's the original author. My guess is the error is there to catch a case where this script is not working e.g. the kernel changed the names of irq threads or we misconfigured the kernel and interrupts are not threaded.

billpittman replied:
Definitely does not need to be an error and I think Gratian instincts about the error being there in the first place are correct.  You could probably make a case for removing it all together seeing as it will be corrected eventually and with this running at startup (and the suggested practice of warm-up iterations), there are multiple opportunities for the threads/IRQs to be scheduled/re-scheduled.

chaitu236 replied:
Changed error to warning and exit 0

chaitu236 committed:
initscripts-nilrt: Add niinitled from os-common

Move nivalidatesystem from os-common into OE as niinitled after removing
unused code

Signed-off-by: Chaitanya Vadrevu <chaitanya.vadrevu@ni.com>

amstewart commented:
@chaitu236 Can you post a V2 summary, so I know what you changed.

chaitu236 commented:
> @chaitu236 Can you post a V2 summary, so I know what you changed.

Addressed all feedback except for the one in nisetupkernelconfig.

Removed mountconfig, nicreatelogdir, xserver-xfce.default
Moved nisetupdump  stuff into nisetupkernelconfig

shruthi-ravi commented on recipes-xfce/xserver-xfce-init/xserver-xfce-init-1.0/xserver-xfce.conf:
```diff
@@ -1,2 +1,10 @@
+	| grep -iq false
+then
+	echo "$0: Disabled in system configuration." >&2
+	exit 0
```
Does this have the effect of exiting from the parent xserver-xfce process too? I assume that is the intent given the message.

gratian replied:
The intent of the original code was to not start the X stack (X server and Xfce desktop environment) if the "Enable Embedded UI" checkbox in MAX is not checked.

shruthi-ravi replied:
Right. Since this file is invoked from [here](https://github.com/ni/meta-nilrt/blob/nilrt/master/dunfell/recipes-xfce/xserver-xfce-init/xserver-xfce-init-1.0/xserver-xfce#L8), I was wondering if this exit would cause the parent process to also finish or would it continue on to line 26, which does not seem to be the intent.

shruthi-ravi replied:
Do you have an update on this @chaitu236?

chaitu236 replied:
Oh sorry I misunderstood the comment. Yes the exit would cause the parent to also exit and it wouldn't continue to line 26.

shruthi-ravi reviewed (commented):
lgtm, pending the two outstanding issues.

chaitu236 commented:
v3 changes: Changed nisetupkernelconfig's error to warning and exit 0
